### ISSUE TITLE
Question: Should IDisposable pattern allow explicit implementation

### ISSUE THREAD

**Issue by Corniel** [SonarSource/sonar-dotnet#2605]
### Description
I'm not sure, but defining `void Dispose()` explicit might be matching the expected implementation. Now it is not supported, but it might be a good idea to do so. What do you think?

### Expected behavior
``` C#
public class Disposable : IDisposable // Should not raise S3881
{
    void IDisposable.Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    /// <summary>Disposes the API test context.</summary>
    protected virtual void Dispose(bool disposing) // Should not raise S2953
    {
        if (!isDisposed && disposing)
        {
            // dispose something.
        }
        isDisposed = true;
    }
    private bool isDisposed;
}
```
### Related information
* SonarC# Version 7.16.0.8981

**Comment by andrei-epure-sonarsource** [comment]
Thanks @Corniel , we need to do some tests to make sure that explicit implementation is handled by the framework

**Comment by andrei-epure-sonarsource** [comment]
So, the following snippet indeed triggers 

```
    public interface IFooBar
    {
        public void Dispose(); // S2953 Methods named "Dispose" should implement "IDisposable.Dispose"
    }

    public class Disposable : IDisposable, IFooBar // S3881 "IDisposable" should be implemented correctly
    {
        void IDisposable.Dispose()
        {
            Console.WriteLine("In IDisposable.Dispose");
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        void IFooBar.Dispose()
        {
            Console.WriteLine("In FooBar");
            throw new NotImplementedException();
        }

        /// <summary>Disposes the API test context.</summary>
        protected virtual void Dispose(bool disposing) // Should not raise S2953
        {
            Console.WriteLine("In virtual Dispose");
            if (!isDisposed && disposing)
            {
                // dispose something.
            }
            isDisposed = true;
        }

        private bool isDisposed;
    }
    class Program
    {

        static void Main(string[] args)
        {

            using (var disposable = new Disposable())
            {
                Console.WriteLine("Main");
            }
        }
    }
```

And the console output is

```
Main
In IDisposable.Dispose
In virtual Dispose
```

Tested with a netcoreapp3.0

However, @Corniel , do you think it's a common scenario to explicitly implement IDisposable.Dispose?

**Comment by andrei-epure-sonarsource** [comment]
Well, the RSPEC clearly says 
>  The class should not implement IDisposable explicitly, e.g. the Dispose() method should be public. 

And looking why that is the case, I found [this SO thread](https://stackoverflow.com/questions/5577252/explicit-implementation-of-idisposable) pointing [to an MSDN article](https://docs.microsoft.com/en-us/archive/msdn-magazine/2007/july/clr-inside-out-digging-into-idisposable)

> Although the using block will work with classes that do have an explicit IDisposable implementation, I recommend that classes never implement the interface this way. If you explicitly implement IDisposable, developers who are exploring your object model using IntelliSense® in Visual Studio® will not notice that the object has a Dispose method and therefore may not take advantage of early cleanup, causing your object to hold onto its resources until the garbage collector runs the finalizer.
